In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDRegressor
from statsmodels.tools.eval_measures import rmse
from scipy.stats import shapiro
from statsmodels.stats.outliers_influence import variance_inflation_factor
from warnings import filterwarnings

In [2]:
df=pd.read_csv("h1n1_vaccine_prediction.csv")

In [3]:
df

unique_id  h1n1_worry  h1n1_awareness  antiviral_medication  \
0              0         1.0             0.0                   0.0   
1              1         3.0             2.0                   0.0   
2              2         1.0             1.0                   0.0   
3              3         1.0             1.0                   0.0   
4              4         2.0             1.0                   0.0   
...          ...         ...             ...                   ...   
26702      26702         2.0             0.0                   0.0   
26703      26703         1.0             2.0                   0.0   
26704      26704         2.0             2.0                   0.0   
26705      26705         1.0             1.0                   0.0   
26706      26706         0.0             0.0                   0.0   

       contact_avoidance  bought_face_mask  wash_hands_frequently  \
0                    0.0               0.0                    0.0   
1                    1.0               0.0                    1.0   
2                    1.0               0.0                    0.0   
3                    1.0               0.0                    1.0   
4                    1.0               0.0                    1.0   
...                  ...               ...                    ...   
26702                1.0               0.0                    0.0   
26703                1.0               0.0                    1.0   
26704                1.0               1.0                    1.0   
26705                0.0               0.0                    0.0   
26706                1.0               0.0                    0.0   

       avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  \
0                         0.0                        1.0               1.0   
1                         0.0                        1.0               1.0   
2                         0.0                        0.0               0.0   
3                         1.0                        0.0               0.0   
4                         1.0                        0.0               1.0   
...                       ...                        ...               ...   
26702                     0.0                        1.0               0.0   
26703                     0.0                        0.0               0.0   
26704                     1.0                        0.0               1.0   
26705                     0.0                        0.0               NaN   
26706                     0.0                        0.0               0.0   

       ...      race     sex               income_level  marital_status  \
0      ...     White  Female              Below Poverty     Not Married   
1      ...     White    Male              Below Poverty     Not Married   
2      ...     White    Male  <= $75,000, Above Poverty     Not Married   
3      ...     White  Female              Below Poverty     Not Married   
4      ...     White  Female  <= $75,000, Above Poverty         Married   
...    ...       ...     ...                        ...             ...   
26702  ...     White  Female  <= $75,000, Above Poverty     Not Married   
26703  ...     White    Male  <= $75,000, Above Poverty     Not Married   
26704  ...     White  Female                        NaN     Not Married   
26705  ...  Hispanic  Female  <= $75,000, Above Poverty         Married   
26706  ...     White    Male  <= $75,000, Above Poverty         Married   

       housing_status          employment                census_msa  \
0                 Own  Not in Labor Force                   Non-MSA   
1                Rent            Employed  MSA, Not Principle  City   
2                 Own            Employed  MSA, Not Principle  City   
3                Rent  Not in Labor Force       MSA, Principle City   
4                 Own            Employed  MSA, Not Principle  City   
...               ...                 ...                       ...   
26

# EDA

In [4]:
df.shape

(26707, 34)

In [5]:
df.head(5)

unique_id  h1n1_worry  h1n1_awareness  antiviral_medication  \
0          0         1.0             0.0                   0.0   
1          1         3.0             2.0                   0.0   
2          2         1.0             1.0                   0.0   
3          3         1.0             1.0                   0.0   
4          4         2.0             1.0                   0.0   

   contact_avoidance  bought_face_mask  wash_hands_frequently  \
0                0.0               0.0                    0.0   
1                1.0               0.0                    1.0   
2                1.0               0.0                    0.0   
3                1.0               0.0                    1.0   
4                1.0               0.0                    1.0   

   avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  ...  \
0                     0.0                        1.0               1.0  ...   
1                     0.0                        1.0               1.0  ...   
2                     0.0                        0.0               0.0  ...   
3                     1.0                        0.0               0.0  ...   
4                     1.0                        0.0               1.0  ...   

    race     sex               income_level  marital_status  housing_status  \
0  White  Female              Below Poverty     Not Married             Own   
1  White    Male              Below Poverty     Not Married            Rent   
2  White    Male  <= $75,000, Above Poverty     Not Married             Own   
3  White  Female              Below Poverty     Not Married            Rent   
4  White  Female  <= $75,000, Above Poverty         Married             Own   

           employment                census_msa  no_of_adults  no_of_children  \
0  Not in Labor Force                   Non-MSA           0.0             0.0   
1            Employed  MSA, Not Principle  City           0.0             0.0   
2            Employed  MSA, Not Principle  City           2.0             0.0   
3  Not in Labor Force       MSA, Principle City           0.0             0.0   
4            Employed  MSA, Not Principle  City           1.0             0.0   

   h1n1_vaccine  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 34 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 34 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   unique_id                  26707 non-null  int64  
 1   h1n1_worry                 26615 non-null  float64
 2   h1n1_awareness             26591 non-null  float64
 3   antiviral_medication       26636 non-null  float64
 4   contact_avoidance          26499 non-null  float64
 5   bought_face_mask           26688 non-null  float64
 6   wash_hands_frequently      26665 non-null  float64
 7   avoid_large_gatherings     26620 non-null  float64
 8   reduced_outside_home_cont  26625 non-null  float64
 9   avoid_touch_face           26579 non-null  float64
 10  dr_recc_h1n1_vacc          24547 non-null  float64
 11  dr_recc_seasonal_vacc      24547 non-null  float64
 12  chronic_medic_condition    25736 non-null  float64
 13  cont_child_undr_6_mnths    25887 non-null  flo

In [7]:
df.describe()

unique_id    h1n1_worry  h1n1_awareness  antiviral_medication  \
count  26707.000000  26615.000000    26591.000000          26636.000000   
mean   13353.000000      1.618486        1.262532              0.048844   
std     7709.791156      0.910311        0.618149              0.215545   
min        0.000000      0.000000        0.000000              0.000000   
25%     6676.500000      1.000000        1.000000              0.000000   
50%    13353.000000      2.000000        1.000000              0.000000   
75%    20029.500000      2.000000        2.000000              0.000000   
max    26706.000000      3.000000        2.000000              1.000000   

       contact_avoidance  bought_face_mask  wash_hands_frequently  \
count       26499.000000      26688.000000           26665.000000   
mean            0.725612          0.068982               0.825614   
std             0.446214          0.253429               0.379448   
min             0.000000          0.000000               0.000000   
25%             0.000000          0.000000               1.000000   
50%             1.000000          0.000000               1.000000   
75%             1.000000          0.000000               1.000000   
max             1.000000          1.000000               1.000000   

       avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  \
count             26620.00000               26625.000000      26579.000000   
mean                  0.35864                   0.337315          0.677264   
std                   0.47961                   0.472802          0.467531   
min                   0.00000                   0.000000          0.000000   
25%                   0.00000                   0.000000          0.000000   
50%                   0.00000                   0.000000          1.000000   
75%                   1.00000                   1.000000          1.000000   
max                   1.00000                   1.000000          1.000000   

       ...  has_health_insur  is_h1n1_vacc_effective  is_h1n1_risky  \
count  ...       14433.00000            26316.000000   26319.000000   
mean   ...           0.87972                3.850623       2.342566   
std    ...           0.32530                1.007436       1.285539   
min    ...           0.00000                1.000000       1.000000   
25%    ...           1.00000                3.000000       1.000000   
50%    ...           1.00000                4.000000       2.000000   
75%    ...           1.00000                5.000000       4.000000   
max    ...           1.00000                5.000000       5.000000   

       sick_from_h1n1_vacc  is_seas_vacc_effective  is_seas_risky  \
count         26312.000000            26245.000000   26193.000000   
mean              2.357670                4.025986       2.719162   
std               1.362766                1.086565       1.385055   
min               1.000000                1.000000       1.000000   
25%               1.000000                4.000000       2.000000   
50%               2.000000                4.000000       2.000000   
75%               4.000000                5.000000       4.000000   
max               5.000000                5.000000       5.000000   

       sick_from_seas_vacc  no_of_adults  no_of_children  h1n1_vaccine  
count         26170.000000  26458.000000    26458.000000  26707.000000  
mean              2.118112      0.886499        0.534583      0.212454  
std               1.332950      0.753422        0.928173      0.409052  
min               1.000000      0.000000        0.000000      0.000000  
25%               1.000000      0.000000        0.000000      0.000000  
50%               2.000000      1.000000        0.000000      0.000000  
75%               4.000000      1.000000        1.000000      0.000000  
max               5.000000      3.000000        3.000000      1.000000  

[8 rows x 25 columns]

# Checking for missing values

In [8]:
df.isnull().sum()

unique_id                        0
h1n1_worry                      92
h1n1_awareness                 116
antiviral_medication            71
contact_avoidance              208
bought_face_mask                19
wash_hands_frequently           42
avoid_large_gatherings          87
reduced_outside_home_cont       82
avoid_touch_face               128
dr_recc_h1n1_vacc             2160
dr_recc_seasonal_vacc         2160
chronic_medic_condition        971
cont_child_undr_6_mnths        820
is_health_worker               804
has_health_insur             12274
is_h1n1_vacc_effective         391
is_h1n1_risky                  388
sick_from_h1n1_vacc            395
is_seas_vacc_effective         462
is_seas_risky                  514
sick_from_seas_vacc            537
age_bracket                      0
qualification                 1407
race                             0
sex                              0
income_level                  4423
marital_status                1408
housing_status      

# There are a lot of missing values

# Treating missing values

In [9]:
mode = df.filter(df.columns).mode()

In [10]:
print(mode)

       unique_id  h1n1_worry  h1n1_awareness  antiviral_medication  \
0              0         2.0             1.0                   0.0   
1              1         NaN             NaN                   NaN   
2              2         NaN             NaN                   NaN   
3              3         NaN             NaN                   NaN   
4              4         NaN             NaN                   NaN   
...          ...         ...             ...                   ...   
26702      26702         NaN             NaN                   NaN   
26703      26703         NaN             NaN                   NaN   
26704      26704         NaN             NaN                   NaN   
26705      26705         NaN             NaN                   NaN   
26706      26706         NaN             NaN                   NaN   

       contact_avoidance  bought_face_mask  wash_hands_frequently  \
0                    1.0               0.0                    1.0   
1                    

In [11]:
cols=df.columns

In [12]:
df[cols]=df[cols].fillna(df.mode().iloc[0])

In [13]:
 mode = df.filter(df.columns).mode()
 cols=df.columns
 df[cols]=df[cols].fillna(df.mode().iloc[0])
    

In [14]:
df.isnull().sum()

unique_id                    0
h1n1_worry                   0
h1n1_awareness               0
antiviral_medication         0
contact_avoidance            0
bought_face_mask             0
wash_hands_frequently        0
avoid_large_gatherings       0
reduced_outside_home_cont    0
avoid_touch_face             0
dr_recc_h1n1_vacc            0
dr_recc_seasonal_vacc        0
chronic_medic_condition      0
cont_child_undr_6_mnths      0
is_health_worker             0
has_health_insur             0
is_h1n1_vacc_effective       0
is_h1n1_risky                0
sick_from_h1n1_vacc          0
is_seas_vacc_effective       0
is_seas_risky                0
sick_from_seas_vacc          0
age_bracket                  0
qualification                0
race                         0
sex                          0
income_level                 0
marital_status               0
housing_status               0
employment                   0
census_msa                   0
no_of_adults                 0
no_of_ch

In [15]:
df

unique_id  h1n1_worry  h1n1_awareness  antiviral_medication  \
0              0         1.0             0.0                   0.0   
1              1         3.0             2.0                   0.0   
2              2         1.0             1.0                   0.0   
3              3         1.0             1.0                   0.0   
4              4         2.0             1.0                   0.0   
...          ...         ...             ...                   ...   
26702      26702         2.0             0.0                   0.0   
26703      26703         1.0             2.0                   0.0   
26704      26704         2.0             2.0                   0.0   
26705      26705         1.0             1.0                   0.0   
26706      26706         0.0             0.0                   0.0   

       contact_avoidance  bought_face_mask  wash_hands_frequently  \
0                    0.0               0.0                    0.0   
1                    1.0               0.0                    1.0   
2                    1.0               0.0                    0.0   
3                    1.0               0.0                    1.0   
4                    1.0               0.0                    1.0   
...                  ...               ...                    ...   
26702                1.0               0.0                    0.0   
26703                1.0               0.0                    1.0   
26704                1.0               1.0                    1.0   
26705                0.0               0.0                    0.0   
26706                1.0               0.0                    0.0   

       avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  \
0                         0.0                        1.0               1.0   
1                         0.0                        1.0               1.0   
2                         0.0                        0.0               0.0   
3                         1.0                        0.0               0.0   
4                         1.0                        0.0               1.0   
...                       ...                        ...               ...   
26702                     0.0                        1.0               0.0   
26703                     0.0                        0.0               0.0   
26704                     1.0                        0.0               1.0   
26705                     0.0                        0.0               1.0   
26706                     0.0                        0.0               0.0   

       ...      race     sex               income_level  marital_status  \
0      ...     White  Female              Below Poverty     Not Married   
1      ...     White    Male              Below Poverty     Not Married   
2      ...     White    Male  <= $75,000, Above Poverty     Not Married   
3      ...     White  Female              Below Poverty     Not Married   
4      ...     White  Female  <= $75,000, Above Poverty         Married   
...    ...       ...     ...                        ...             ...   
26702  ...     White  Female  <= $75,000, Above Poverty     Not Married   
26703  ...     White    Male  <= $75,000, Above Poverty     Not Married   
26704  ...     White  Female  <= $75,000, Above Poverty     Not Married   
26705  ...  Hispanic  Female  <= $75,000, Above Poverty         Married   
26706  ...     White    Male  <= $75,000, Above Poverty         Married   

       housing_status          employment                census_msa  \
0                 Own  Not in Labor Force                   Non-MSA   
1                Rent            Employed  MSA, Not Principle  City   
2                 Own            Employed  MSA, Not Principle  City   
3                Rent  Not in Labor Force       MSA, Principle City   
4                 Own            Employed  MSA, Not Principle  City   
...               ...                 ...                       ...   
26

In [16]:
df.drop(["unique_id"],axis=1,inplace=True)

# Encoding

In [17]:
df.dtypes

h1n1_worry                   float64
h1n1_awareness               float64
antiviral_medication         float64
contact_avoidance            float64
bought_face_mask             float64
wash_hands_frequently        float64
avoid_large_gatherings       float64
reduced_outside_home_cont    float64
avoid_touch_face             float64
dr_recc_h1n1_vacc            float64
dr_recc_seasonal_vacc        float64
chronic_medic_condition      float64
cont_child_undr_6_mnths      float64
is_health_worker             float64
has_health_insur             float64
is_h1n1_vacc_effective       float64
is_h1n1_risky                float64
sick_from_h1n1_vacc          float64
is_seas_vacc_effective       float64
is_seas_risky                float64
sick_from_seas_vacc          float64
age_bracket                   object
qualification                 object
race                          object
sex                           object
income_level                  object
marital_status                object
h

In [19]:
df=pd.get_dummies(df,columns=['housing_status'])

In [20]:
df.head()

h1n1_worry  h1n1_awareness  antiviral_medication  contact_avoidance  \
0         1.0             0.0                   0.0                0.0   
1         3.0             2.0                   0.0                1.0   
2         1.0             1.0                   0.0                1.0   
3         1.0             1.0                   0.0                1.0   
4         2.0             1.0                   0.0                1.0   

   bought_face_mask  wash_hands_frequently  avoid_large_gatherings  \
0               0.0                    0.0                     0.0   
1               0.0                    1.0                     0.0   
2               0.0                    0.0                     0.0   
3               0.0                    1.0                     1.0   
4               0.0                    1.0                     1.0   

   reduced_outside_home_cont  avoid_touch_face  dr_recc_h1n1_vacc  ...  \
0                        1.0               1.0                0.0  ...   
1                        1.0               1.0                0.0  ...   
2                        0.0               0.0                0.0  ...   
3                        0.0               0.0                0.0  ...   
4                        0.0               1.0                0.0  ...   

   marital_status_Married  marital_status_Not Married  employment_Employed  \
0                       0                           1                    0   
1                       0                           1                    1   
2                       0                           1                    1   
3                       0                           1                    0   
4                       1                           0                    1   

   employment_Not in Labor Force  employment_Unemployed  \
0                              1                      0   
1                              0                      0   
2                              0                      0   
3                              1                      0   
4                              0                      0   

   census_msa_MSA, Not Principle  City  census_msa_MSA, Principle City  \
0                                    0                               0   
1                                    1                               0   
2                                    1                               0   
3                                    0                               1   
4                                    1                               0   

   census_msa_Non-MSA  housing_status_Own  housing_status_Rent  
0                   1                   1                    0  
1                   0                   0                    1  
2                   0                   1                    0  
3                   0                   0                    1  
4                   0                   1                    0  

[5 rows x 52 columns]

# Building Models

## 1.SGD

In [26]:
from sklearn.linear_model import SGDClassifier
x=df.drop(["h1n1_vaccine"], axis = 1)
y=df["h1n1_vaccine"]
x_train,x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)

In [27]:
sgd = SGDClassifier(loss="hinge", penalty="l2", max_iter=5)

In [28]:
sgd.fit(x_train, y_train)


C:\Users\sanke\anaconda3\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:574: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "


SGDClassifier(max_iter=5)

In [29]:
sgd.score(x_train,y_train)

0.7400235369637317

In [30]:
sgd.score(x_test,y_test)

0.7373018844377885

### Checking if score can be imporoved using preprocessing scaling

### Standard scaling

In [31]:
from sklearn.preprocessing import StandardScaler    
scaler = StandardScaler()
scaler.fit(x_train)
x_train= scaler.fit_transform(x_train)    
x_test= scaler.transform(x_test) 

In [32]:
sgd.fit(x_train, y_train)


C:\Users\sanke\anaconda3\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:574: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "


SGDClassifier(max_iter=5)

In [33]:
sgd.score(x_train,y_train)

0.7925537605648871

In [34]:
sgd.score(x_test,y_test)

0.7860975914139523

#### We can se that there is 5% improvement in the score when standard scaling is used

# 2.Logistic Regression

In [37]:
from sklearn.linear_model import LogisticRegression

model=LogisticRegression()
model.fit(x_train,y_train)

LogisticRegression()

In [39]:
predictions=model.predict(x_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test,predictions)

0.8308997878447523

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 52 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   h1n1_worry                              26707 non-null  float64
 1   h1n1_awareness                          26707 non-null  float64
 2   antiviral_medication                    26707 non-null  float64
 3   contact_avoidance                       26707 non-null  float64
 4   bought_face_mask                        26707 non-null  float64
 5   wash_hands_frequently                   26707 non-null  float64
 6   avoid_large_gatherings                  26707 non-null  float64
 7   reduced_outside_home_cont               26707 non-null  float64
 8   avoid_touch_face                        26707 non-null  float64
 9   dr_recc_h1n1_vacc                       26707 non-null  float64
 10  dr_recc_seasonal_vacc                   26707 non-null  fl

# 3.Decision Tree

In [46]:
from sklearn.tree import DecisionTreeClassifier
dTree=DecisionTreeClassifier(criterion="gini",random_state=0,max_depth=5)
dTree.fit(x_train,y_train)
dTree.score(x_train,y_train)
dTree.score(x_test,y_test)

0.8262822912766754

In [47]:
dTree.score(x_train,y_train)


0.8347063228843479

# Result

#### SGD
Train Score:0.7925537605648871
  Test Score:0.7860975914139523

#### Logistic Regression
Accuracy Score:0.8308997878447523

#### Decision Tree
Train_Score:0.8347063228843479
    Test_Score:0.8262822912766754